File: pkg/api/totp.go
// Copyright 2020 Matt Montgomery
// SPDX-License-Identifier: AGPL-3.0-or-later

package api

import (
	"net/http"
	"log"

	"github.com/ConfusedPolarBear/lifeguard/pkg/config"
	"github.com/ConfusedPolarBear/lifeguard/pkg/zpool"
	
	"github.com/gorilla/mux"
)

func SetupTOTP(r *mux.Router) {
	r.HandleFunc("/api/v0/tfa/totp/initialize", initializeHandler).Methods("GET")
	r.HandleFunc("/api/v0/tfa/totp/save", saveHandler).Methods("POST")
	r.HandleFunc("/api/v0/tfa/totp/authenticate", challengeHandler).Methods("POST")
}

func initializeHandler(w http.ResponseWriter, r *http.Request) {
	username := getUsernameQuiet(r)
	if username == "" {
		return
	}

	secret, image := config.InitializeTOTP(username)
	image = "data:image/png;base64," + image

	ret := struct {
		Secret string
		Image string
	} {
		secret,
		image,
	}

	w.Write(zpool.Encode(ret))
}

func saveHandler(w http.ResponseWriter, r *http.Request) {
	username := getUsernameQuiet(r)
	if username == "" {
		return
	}

	secret, secretOk := GetParameter(r, "secret")
	if !secretOk {
		ReportMissing(w)
		return
	}

	code, codeOk := GetParameter(r, "code")
	if !codeOk {
		ReportMissing(w)
		return
	}

	ok := config.SaveTOTP(username, secret, code)
	if !ok {
		http.Error(w, "Invalid code - check that the time is in sync on the server and your phone", http.StatusBadRequest)
		return
	}

	http.Error(w, "OK", http.StatusOK)
}

func challengeHandler(w http.ResponseWriter, r *http.Request) {
	session := getSession(r)

	username := getPartialAuth(r)
	if username == "" {
		return
	}

	code, codeOk := GetParameter(r, "code")
	if !codeOk {
		ReportMissing(w)
		return
	}

	ok := config.VerifyTOTP(username, code)
	if !ok {
		http.Error(w, "Invalid code", http.StatusForbidden)
		return
	}

	session.Values["authenticated"] = true
	err := session.Save(r, w)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		log.Printf("Unable to save session: %s", err)
		return
	}

	http.Error(w, "OK", http.StatusOK)
}

File: web/src/api/totp.js
// Copyright 2020 Matt Montgomery
// SPDX-License-Identifier: AGPL-3.0-or-later

import * as ApiClient from '../apiClient.js';

export async function Initialize() {
	const res = await fetch('/api/v0/tfa/totp/initialize');
	return await res.json();
}

export async function IsEnabled() {
	const res = await fetch('/api/v0/tfa/enabled');
	let json = await res.json();
	return json.Enabled;
}

export async function Save(secret, code) {
	const res = await ApiClient.Post('/api/v0/tfa/totp/save', {
		secret: secret,
		code: code
	});

	if (!res.ok) {
		return Promise.reject(await res.text());
	}

	return Promise.resolve(true);
}

export async function Authenticate(code) {
	const res = await ApiClient.Post('/api/v0/tfa/totp/authenticate', {
		code: code
	});

	return Promise.resolve(res.ok);
}
